diff --git a/lang/node/Makefile b/lang/node/Makefile
index 9df8876..164585d 100644
--- a/lang/node/Makefile
+++ b/lang/node/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2006-2011 OpenWrt.org
+# Copyright (C) 2006-2016 OpenWrt.org
 #
 # This is free software, licensed under the GNU General Public License v2.
 # See /LICENSE for more information.
@@ -8,8 +8,8 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=node
-PKG_VERSION:=v0.12.7
-PKG_RELEASE:=2
+PKG_VERSION:=v0.12.13
+PKG_RELEASE:=1
 
 PKG_SOURCE:=node-$(PKG_VERSION).tar.gz
 PKG_SOURCE_URL:=http://nodejs.org/dist/${PKG_VERSION}
@@ -51,12 +51,14 @@ CONFIGURE_ARGS= \
 	--without-snapshot \
 	--shared-zlib \
 	--shared-openssl \
+	--with-mips-float-abi=soft \
 	--prefix=/usr
 
 HOST_CONFIGURE_VARS:=
 HOST_CONFIGURE_ARGS:= \
 	--dest-os=linux \
 	--without-snapshot \
+	--with-mips-float-abi=soft \
 	--prefix=$(STAGING_DIR)/host/
 
 HOST_CONFIGURE_CMD:=python ./configure
diff --git a/utils/rng-tools/files/rngd.init b/utils/rng-tools/files/rngd.init
index 805559c..33a58f5 100644
--- a/utils/rng-tools/files/rngd.init
+++ b/utils/rng-tools/files/rngd.init
@@ -1,7 +1,7 @@
 #!/bin/sh /etc/rc.common
 # Copyright (C) 2011-2014 OpenWrt.org
 
-START=98
+START=48
 
 RNGD_INTERVAL=30
 RNGD_AMOUNT=4000
