diff --git a/target/linux/ar71xx/Makefile b/target/linux/ar71xx/Makefile
index 594b48e..c0d867c 100644
--- a/target/linux/ar71xx/Makefile
+++ b/target/linux/ar71xx/Makefile
@@ -10,7 +10,7 @@ ARCH:=mips
 BOARD:=ar71xx
 BOARDNAME:=Atheros AR7xxx/AR9xxx
 FEATURES:=mips16
-CPU_TYPE=34kc
+CPU_TYPE=24kc
 SUBTARGETS:=generic nand mikrotik
 
 KERNEL_PATCHVER:=4.4
diff --git a/target/linux/ar71xx/base-files/etc/board.d/01_leds b/target/linux/ar71xx/base-files/etc/board.d/01_leds
index 6936b8c..d89eae7 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/01_leds
+++ b/target/linux/ar71xx/base-files/etc/board.d/01_leds
@@ -55,6 +55,11 @@ arduino-yun)
 	ucidef_set_led_usbdev "usb" "USB" "arduino:white:usb" "1-1.1"
 	;;
 
+yun-shield)
+	ucidef_set_led_netdev "wan" "WAN" "yun-shield:orange:wan" "eth1"
+	ucidef_set_led_wlan "wlan" "WLAN" "yun-shield:blue:wlan" "phy0tpt"
+	;;
+
 ap113)
 	ucidef_set_led_usbdev "usb" "USB" "ap113:green:usb" "1-1"
 	;;
diff --git a/target/linux/ar71xx/base-files/etc/board.d/02_network b/target/linux/ar71xx/base-files/etc/board.d/02_network
index 1ed593c..c0ac606 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/02_network
+++ b/target/linux/ar71xx/base-files/etc/board.d/02_network
@@ -382,6 +382,7 @@ cap324)
 	;;
 
 arduino-yun |\
+yun-shield |\
 tl-wa801nd-v3 |\
 dir-505-a1)
 	ucidef_set_interface_lan "eth1"
diff --git a/target/linux/ar71xx/base-files/lib/ar71xx.sh b/target/linux/ar71xx/base-files/lib/ar71xx.sh
index 5b81f73..86fa5ac 100755
--- a/target/linux/ar71xx/base-files/lib/ar71xx.sh
+++ b/target/linux/ar71xx/base-files/lib/ar71xx.sh
@@ -386,6 +386,9 @@ ar71xx_board_detect() {
 	*"Arduino Yun")
 		name="arduino-yun"
 		;;
+	*"Yun Shield")
+		name="yun-shield"
+		;;
 	*AP113)
 		name="ap113"
 		;;
diff --git a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
index 313e2e6..2ac4abb 100755
--- a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
@@ -181,6 +181,7 @@ platform_check_image() {
 	alfa-ap96 | \
 	alfa-nx | \
 	arduino-yun | \
+	yun-shield | \
 	ap113 | \
 	ap121 | \
 	ap121-mini | \
diff --git a/target/linux/ar71xx/config-4.1 b/target/linux/ar71xx/config-4.1
index 5939891..0f7657d 100644
--- a/target/linux/ar71xx/config-4.1
+++ b/target/linux/ar71xx/config-4.1
@@ -49,6 +49,7 @@ CONFIG_ATH79_MACH_AP83=y
 CONFIG_ATH79_MACH_AP96=y
 CONFIG_ATH79_MACH_ARCHER_C7=y
 CONFIG_ATH79_MACH_ARDUINO_YUN=y
+CONFIG_ATH79_MACH_ARDUINO_YUN_SHIELD=y
 CONFIG_ATH79_MACH_AW_NR580=y
 CONFIG_ATH79_MACH_BHU_BXU2000N2_A=y
 CONFIG_ATH79_MACH_BSB=y
@@ -296,6 +297,7 @@ CONFIG_MARVELL_PHY=y
 CONFIG_MDIO_BOARDINFO=y
 CONFIG_MICREL_PHY=y
 CONFIG_MIPS=y
+CONFIG_MIPS_FPU_EMULATOR=y
 # CONFIG_MIPS_HUGE_TLB_SUPPORT is not set
 CONFIG_MIPS_L1_CACHE_SHIFT=5
 CONFIG_MIPS_MACHINE=y
diff --git a/target/linux/ar71xx/config-4.4 b/target/linux/ar71xx/config-4.4
index 79d703b..8559f49 100644
--- a/target/linux/ar71xx/config-4.4
+++ b/target/linux/ar71xx/config-4.4
@@ -52,6 +52,7 @@ CONFIG_ATH79_MACH_AP83=y
 CONFIG_ATH79_MACH_AP96=y
 CONFIG_ATH79_MACH_ARCHER_C7=y
 CONFIG_ATH79_MACH_ARDUINO_YUN=y
+CONFIG_ATH79_MACH_ARDUINO_YUN_SHIELD=y
 CONFIG_ATH79_MACH_AW_NR580=y
 CONFIG_ATH79_MACH_BHU_BXU2000N2_A=y
 CONFIG_ATH79_MACH_BSB=y
@@ -321,6 +322,7 @@ CONFIG_MIPS_CLOCK_VSYSCALL=y
 CONFIG_MIPS_CMDLINE_FROM_BOOTLOADER=y
 # CONFIG_MIPS_CMDLINE_FROM_DTB is not set
 # CONFIG_MIPS_ELF_APPENDED_DTB is not set
+CONFIG_MIPS_FPU_EMULATOR=y
 # CONFIG_MIPS_HUGE_TLB_SUPPORT is not set
 CONFIG_MIPS_L1_CACHE_SHIFT=5
 CONFIG_MIPS_MACHINE=y
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt b/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
index 52b9b55..f8a5a7a 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
+++ b/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
@@ -103,6 +103,19 @@ config ATH79_MACH_ARDUINO_YUN
 	  Say 'Y' here if you want your kernel to support the
 	  Arduino Yun.
 
+config ATH79_MACH_ARDUINO_YUN_SHIELD
+	bool "Arduino/Genuino Yun Shield"
+	select SOC_AR933X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+	help
+	  Say 'Y' here if you want your kernel to support the
+	  Arduino/Genuino Yun Shield.
+
 config ATH79_MACH_AP113
 	bool "Atheros AP113 board support"
 	select SOC_AR724X
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/Makefile b/target/linux/ar71xx/files/arch/mips/ath79/Makefile
index f8b0b2b..413fa32 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/Makefile
+++ b/target/linux/ar71xx/files/arch/mips/ath79/Makefile
@@ -47,6 +47,7 @@ obj-$(CONFIG_ATH79_MACH_ANTMINER_S1)+= mach-antminer-s1.o
 obj-$(CONFIG_ATH79_MACH_ANTMINER_S3)+= mach-antminer-s3.o
 obj-$(CONFIG_ATH79_MACH_ANTROUTER_R1)  += mach-antrouter-r1.o
 obj-$(CONFIG_ATH79_MACH_ARDUINO_YUN)	+= mach-arduino-yun.o
+obj-$(CONFIG_ATH79_MACH_ARDUINO_YUN_SHIELD)	+= mach-arduino-yun-shield.o
 obj-$(CONFIG_ATH79_MACH_AP113)		+= mach-ap113.o
 obj-$(CONFIG_ATH79_MACH_AP121)		+= mach-ap121.o
 obj-$(CONFIG_ATH79_MACH_AP132)		+= mach-ap132.o
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-arduino-yun-shield.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-arduino-yun-shield.c
index e69de29..8d74fed 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/mach-arduino-yun-shield.c
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-arduino-yun-shield.c
@@ -0,0 +1,147 @@
+/*
+ *  Yun Shield support
+ *
+ *  Copyright (C) 2011-2012 Gabor Juhos <juhosg@openwrt.org>
+ *
+ *  This program is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 as published
+ *  by the Free Software Foundation.
+ */
+
+#include "dev-eth.h"
+#include "dev-gpio-buttons.h"
+#include "dev-leds-gpio.h"
+#include "dev-m25p80.h"
+#include "dev-spi.h"
+#include "dev-usb.h"
+#include "dev-wmac.h"
+#include "machtypes.h"
+#include <asm/mach-ath79/ar71xx_regs.h>
+#include <asm/mach-ath79/ath79.h>
+#include <linux/gpio.h>
+#include "common.h"
+
+#define DS_GPIO_LED_READY		13
+#define DS_GPIO_LED_WAN		1
+#define DS_GPIO_LED_WLAN		0
+#define DS_GPIO_LED_USER		15
+
+#define DS_GPIO_OE			22 /*SPI*/
+#define DS_GPIO_AVR_RESET		12
+
+#define DS_GPIO_OE2			16 /*HANDSHAKE */
+#define DS_GPIO_UART_ENA		17
+#define DS_GPIO_CONF_BTN		11
+
+#define DS_KEYS_POLL_INTERVAL		20	/* msecs */
+#define DS_KEYS_DEBOUNCE_INTERVAL	(3 * DS_KEYS_POLL_INTERVAL)
+
+#define DS_MAC0_OFFSET			0x0000
+#define DS_MAC1_OFFSET			0x0006
+#define DS_CALDATA_OFFSET		0x1000
+#define DS_WMAC_MAC_OFFSET		0x1002
+
+static struct gpio_led ds_leds_gpio[] __initdata = {
+	{
+		.name		= "yun-shield:white:system",
+		.gpio		= DS_GPIO_LED_READY,
+		.active_low	= 1,
+	},
+	{
+		.name		= "yun-shield:orange:wan",
+		.gpio		= DS_GPIO_LED_WAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "yun-shield:blue:wlan",
+		.gpio		= DS_GPIO_LED_WLAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "yun-shield:red:user",
+		.gpio		= DS_GPIO_LED_USER,
+		.active_low	= 0,
+	},
+};
+
+static struct gpio_keys_button ds_gpio_keys[] __initdata = {
+	{
+		.desc		= "configuration button",
+		.type		= EV_KEY,
+		.code		= KEY_WPS_BUTTON,
+		.debounce_interval = DS_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= DS_GPIO_CONF_BTN,
+		.active_low	= 1,
+	},
+};
+
+static void __init ds_common_setup(void)
+{
+	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
+	ath79_register_m25p80(NULL);
+
+	ath79_register_wmac(art + DS_CALDATA_OFFSET,
+			    art + DS_WMAC_MAC_OFFSET);
+	ath79_init_mac(ath79_eth0_data.mac_addr, art + DS_MAC0_OFFSET, 0);
+	ath79_init_mac(ath79_eth1_data.mac_addr, art + DS_MAC1_OFFSET, 0);
+
+	ath79_register_mdio(0, 0x0);
+
+	/* LAN ports */
+	ath79_register_eth(0);
+
+	/* WAN port */
+	ath79_register_eth(1);
+}
+
+static void __init ds_setup(void)
+{
+	u32 t;
+
+	ds_common_setup();
+
+	ath79_register_leds_gpio(-1, ARRAY_SIZE(ds_leds_gpio),
+				 ds_leds_gpio);
+	ath79_register_gpio_keys_polled(-1, DS_KEYS_POLL_INTERVAL,
+					ARRAY_SIZE(ds_gpio_keys),
+					ds_gpio_keys);
+	ath79_register_usb();
+
+	/* use the swtich_led directly form sysfs */
+	ath79_gpio_function_disable(AR933X_GPIO_FUNC_ETH_SWITCH_LED0_EN |
+				    AR933X_GPIO_FUNC_ETH_SWITCH_LED1_EN |
+				    AR933X_GPIO_FUNC_ETH_SWITCH_LED2_EN |
+				    AR933X_GPIO_FUNC_ETH_SWITCH_LED3_EN |
+				    AR933X_GPIO_FUNC_ETH_SWITCH_LED4_EN);
+
+	/*
+	 * Disable the Function for some pins to have GPIO functionality active
+	 * GPIO6-7-8 and GPIO11
+	 */
+	ath79_gpio_function_setup(AR933X_GPIO_FUNC_JTAG_DISABLE | AR933X_GPIO_FUNC_I2S_MCK_EN, 0);
+	ath79_gpio_function2_setup(AR933X_GPIO_FUNC2_JUMPSTART_DISABLE | AR933X_GPIO_FUNC2_WPS_DISABLE, 0);
+
+	printk("Setting GPIO\n");
+
+	/* Enable GPIO26 instead of MDC function */
+	t = ath79_reset_rr(AR933X_RESET_REG_BOOTSTRAP);
+	t |= AR933X_BOOTSTRAP_MDIO_GPIO_EN;
+	ath79_reset_wr(AR933X_RESET_REG_BOOTSTRAP, t);
+
+	/* enable OE of level shifter */
+	if (gpio_request_one(DS_GPIO_OE,
+	    GPIOF_OUT_INIT_LOW | GPIOF_EXPORT_DIR_FIXED, "OE-1") != 0)
+		printk("Error setting GPIO OE\n");
+
+	/* enable uart */
+	if (gpio_request_one(DS_GPIO_UART_ENA,
+	    GPIOF_OUT_INIT_HIGH | GPIOF_EXPORT_DIR_FIXED, "UART-ENA") != 0)
+		printk("Error setting GPIO Uart Enable\n");
+
+	/* enable OE2 of level shifter */
+	if (gpio_request_one(DS_GPIO_OE2,
+	    GPIOF_OUT_INIT_LOW | GPIOF_EXPORT_DIR_FIXED, "OE-2") != 0)
+		printk("Error setting GPIO OE2\n");
+}
+
+MIPS_MACHINE(ATH79_MACH_ARDUINO_YUN_SHIELD, "Yun_Shield", "Yun Shield", ds_setup);
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h b/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
index f40b4a4..55c9e22 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
+++ b/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
@@ -27,6 +27,7 @@ enum ath79_mach_type {
 	ATH79_MACH_ANTMINER_S3,		/* Antminer S3 */
 	ATH79_MACH_ANTROUTER_R1,	/* Antrouter R1 */
 	ATH79_MACH_ARDUINO_YUN,		/* Yun */
+	ATH79_MACH_ARDUINO_YUN_SHIELD,	/* Yun Shield */
 	ATH79_MACH_AP113,		/* Atheros AP113 reference board */
 	ATH79_MACH_AP121,		/* Atheros AP121 reference board */
 	ATH79_MACH_AP121_MINI,		/* Atheros AP121-MINI reference board */
diff --git a/target/linux/ar71xx/generic/profiles/arduino.mk b/target/linux/ar71xx/generic/profiles/arduino.mk
index 8ae5054..1122030 100644
--- a/target/linux/ar71xx/generic/profiles/arduino.mk
+++ b/target/linux/ar71xx/generic/profiles/arduino.mk
@@ -17,3 +17,24 @@ endef
 
 $(eval $(call Profile,Yun))
 
+define Profile/YUNSHIELD
+	NAME:=Arduino/Genuino Yun Shield
+	PACKAGES:=kmod-usb-core kmod-usb2
+endef
+
+define Profile/YUNSHIELD/Description
+	Package set optimized for the Arduino/Genuino Yun Shield.
+endef
+
+$(eval $(call Profile,YUNSHIELD))
+
+define Profile/YUNCOMPAT
+	NAME:=Arduino/Genuino Yun compatible boards
+	PACKAGES:=kmod-usb-core kmod-usb2
+endef
+
+define Profile/YUNCOMPAT/Description
+	Packages for the Arduino/Genuino Yun, Yun Shield.
+endef
+
+$(eval $(call Profile,YUNCOMPAT))
diff --git a/target/linux/ar71xx/image/legacy.mk b/target/linux/ar71xx/image/legacy.mk
index 1d8898a..fc7eb00 100644
--- a/target/linux/ar71xx/image/legacy.mk
+++ b/target/linux/ar71xx/image/legacy.mk
@@ -1013,6 +1013,7 @@ $(eval $(call SingleProfile,AthLzma,64k,WPJ558_16M,wpj558-16M,WPJ558,ttyS0,11520
 $(eval $(call SingleProfile,AthLzma,64k,WRTNODE2Q,wrtnode2q,WRTNODE2Q,ttyS0,115200,$$(wrtnode2q_mtdlayout),KRuImage))
 $(eval $(call SingleProfile,AthLzma,64k,YUN_8M,yun-8M,Yun,ttyATH0,250000,$$(yun_mtdlayout_8M),RKuImage))
 $(eval $(call SingleProfile,AthLzma,64k,YUN_16M,yun-16M,Yun,ttyATH0,250000,$$(yun_mtdlayout_16M),RKuImage))
+$(eval $(call SingleProfile,AthLzma,64k,YUNSHIELD,yun-shield,Yun_Shield,ttyATH0,250000,$$(yun_mtdlayout_16M),RKuImage))
 
 $(eval $(call SingleProfile,Belkin,64k,F9K1115V2,f9k1115v2,F9K1115V2,ttyS0,115200,$$(f9k1115v2_mtdlayout),BR-6679BAC))
 
@@ -1141,6 +1142,7 @@ $(eval $(call MultiProfile,WPJ344,WPJ344_16M))
 $(eval $(call MultiProfile,WPJ531,WPJ531_16M))
 $(eval $(call MultiProfile,WPJ558,WPJ558_16M))
 $(eval $(call MultiProfile,Yun,YUN_16M YUN_8M))
+$(eval $(call MultiProfile,YUNCOMPAT,YUN_16M YUN_8M YUNSHIELD))
 
 $(eval $(call MultiProfile,Minimal,$(SINGLE_PROFILES)))
 $(eval $(call MultiProfile,Madwifi,EAP7660D WP543))
diff --git a/target/linux/ar71xx/patches-4.1/820-MIPS-ath79-add_gpio_function2_setup.patch b/target/linux/ar71xx/patches-4.1/820-MIPS-ath79-add_gpio_function2_setup.patch
index 5c90057..ef58e80 100644
--- a/target/linux/ar71xx/patches-4.1/820-MIPS-ath79-add_gpio_function2_setup.patch
+++ b/target/linux/ar71xx/patches-4.1/820-MIPS-ath79-add_gpio_function2_setup.patch
@@ -61,10 +61,11 @@ functions on the Arduino Yun.
  
  #define AR934X_GPIO_REG_OUT_FUNC0	0x2c
  #define AR934X_GPIO_REG_OUT_FUNC1	0x30
-@@ -971,6 +972,8 @@
+@@ -971,6 +972,9 @@
  #define AR724X_GPIO_FUNC_UART_EN		BIT(1)
  #define AR724X_GPIO_FUNC_JTAG_DISABLE		BIT(0)
  
++#define AR933X_GPIO_FUNC2_WPS_DISABLE		BIT(8)
 +#define AR933X_GPIO_FUNC2_JUMPSTART_DISABLE 	BIT(9)
 +
  #define AR913X_GPIO_FUNC_WMAC_LED_EN		BIT(22)
diff --git a/target/linux/ar71xx/patches-4.4/820-MIPS-ath79-add_gpio_function2_setup.patch b/target/linux/ar71xx/patches-4.4/820-MIPS-ath79-add_gpio_function2_setup.patch
index f7f6b72..ebf84d5 100644
--- a/target/linux/ar71xx/patches-4.4/820-MIPS-ath79-add_gpio_function2_setup.patch
+++ b/target/linux/ar71xx/patches-4.4/820-MIPS-ath79-add_gpio_function2_setup.patch
@@ -56,10 +56,11 @@ functions on the Arduino Yun.
  
  #define AR934X_GPIO_REG_OUT_FUNC0	0x2c
  #define AR934X_GPIO_REG_OUT_FUNC1	0x30
-@@ -971,6 +972,8 @@
+@@ -971,6 +972,9 @@
  #define AR724X_GPIO_FUNC_UART_EN		BIT(1)
  #define AR724X_GPIO_FUNC_JTAG_DISABLE		BIT(0)
  
++#define AR933X_GPIO_FUNC2_WPS_DISABLE		BIT(8)
 +#define AR933X_GPIO_FUNC2_JUMPSTART_DISABLE 	BIT(9)
 +
  #define AR913X_GPIO_FUNC_WMAC_LED_EN		BIT(22)
